### ISSUE TITLE
tailscaled failing to start on GKE Autopilot cluster in userspace-networking

### ISSUE THREAD

**Issue by BrianAdams** [tailscale/tailscale#2822]
Simply trying to run the container on [Google GKE Autopilot ](https://cloud.google.com/kubernetes-engine/docs/concepts/autopilot-overview)

Turned on all the capabilities that are allowed:
```
        securityContext:
          capabilities:
            add: ["NET_BIND_SERVICE","SETPCAP","MKNOD","CHOWN","DAC_OVERRIDE","FOWNER","FSETID","KILL","SETGID","SETUID","SYS_CHROOT","SETFCAP"]      
```

GKE Autopilot does drop the NET_RAW capability after start up if that is at all relevant.

But when tailscaled runs it fails with the unable to bind udp4 port 0: ... SO_BINDTODEVICE: operation not permitted.

I am using the[ google cloud run instructions](https://tailscale.com/kb/1108/cloudrun/) to create the Docker container.

Tried with both 
`--tun=userspace-networking --socks5-server=localhost:1055`
and
`--tun=userspace-networking`

```
failed to connect to local tailscaled (which appears to be running). Got error: Get "http://local-tailscaled.sock/localapi/v0/status": dial unix /var/run/tailscale/tailscaled.sock: connect: no such file or directory
2021/09/09 05:10:25 logtail started
2021/09/09 05:10:25 Program starting: v1.14.0-t5cff36945-g809e87bba, Go 1.17-tsdd6b269c77: []string{"/app/tailscaled", "--tun=userspace-networking"}
2021/09/09 05:10:25 LogID: f045a2528d8f0acedd93da4d1d970a3f252997628654077d393bd2d4d1cc30c9
2021/09/09 05:10:25 logpolicy: using system state directory "/var/lib/tailscale"
logpolicy.Read /var/lib/tailscale/tailscaled.log.conf: open /var/lib/tailscale/tailscaled.log.conf: no such file or directory
2021/09/09 05:10:25 wgengine.NewUserspaceEngine(tun "userspace-networking") ...
2021/09/09 05:10:25 dns: using dns.noopManager
2021/09/09 05:10:25 link state: interfaces.State{defaultRoute=eth0 ifs={eth0:[10.52.0.155/26]} v4=true v6=false}
2021/09/09 05:10:25 magicsock: unable to bind udp4 port 0: listen udp4 :0: setting SO_BINDTODEVICE: operation not permitted
2021/09/09 05:10:25 wgengine.NewUserspaceEngine(tun "userspace-networking") error: wgengine: magicsock: initialBind IPv4 failed: failed to bind any ports (tried [0])
2021/09/09 05:10:25 wgengine.New: wgengine: magicsock: initialBind IPv4 failed: failed to bind any ports (tried [0])
2021/09/09 05:10:25 flushing log.
```

Not sure where to look next. Any ideas?

**Comment by DentonGentry** [comment]
The problem being run into is in https://github.com/tailscale/tailscale/blob/main/net/netns/netns_linux.go
It is important that sockets which will be used to contact Tailscale infrastructure, like the coordination server or DERP relays, use the underlying Internet connection to do so not try to go over tailscale0. Trying to use tailscale to run tailscale might sortof work, but is mostly unable to recover from any error and never works well.

netns will first try using FW marks (requires CAP_NET_ADMIN) and falls back to SO_BINDTODEVICE (requires CAP_NET_RAW). Without either one, it is stuck.

In 3/2020 https://patchwork.ozlabs.org/project/netdev/patch/20200331132009.1306283-1-vincent@bernat.ch/ went into Linux, which will allow SO_BINDTODEVICE without CAP_NET_RAW. I would think this would have made it into a Linux release by now, but don't specifically know. If you have an option to use a very recent kernel, or apply that change and build your own kernel, that might allow SO_BINDTODEVICE in the GKE cluster.

**Comment by BrianAdams** [comment]
@DentonGentry Very helpful response. Thanks.

Looks like it is possible to override the drop of the CAP_NET_RAW in GKE autopilot, and that indeed allowed the tailscaled process to start.

```
        securityContext:
          capabilities:
            add: ["NET_RAW","NET_BIND_SERVICE","SETPCAP","MKNOD","CHOWN","DAC_OVERRIDE","FOWNER","FSETID","KILL","SETGID","SETUID","SYS_CHROOT","SETFCAP"]
            drop: []
```

I'll go through and ensure I can remove the rest of the capabilities, but for the record it started with the above set.

**Comment by DentonGentry** [comment]
Since the main issue appears to be resolved I'm going to close this, feel free to open other issues if something else comes up.

**Comment by DentonGentry** [comment]
We've realized that we can relax the need for CAP_NET_RAW. SO_BINDTODEVICE is needed because we don't want Tailscale control packets to egress from the tailscale0 interface. However in --tun=userspace-networking mode, there is no tailscale0 interface. There is no reason to bind to a device, so we'll just skip doing so.

I think this should be fixed in 1.15.111, allowing tailscale to work in GKE even without CAP_NET_RAW. Please reopen the bug if not. https://pkgs.tailscale.com/unstable/
